### PR TITLE
Universal plugins [wip]

### PR DESCRIPTION
**Pull request by RandomShaper**
## Goal

The main idea is creating a system that allows plugins to be installed into a specific project or globally (at the OS user level) so they are available for every project.

## Remarks

- Each commit adds a logical step on top of which others are built.
- Given the big scope of this PR, it's better to review it on a per-commit basis.
- Some of them (especially, the one marked with `[f]`) are meant to be squashed together before merging.
- Feedback and testing is welcome!

### Specification

Please read the full [design document](https://docs.google.com/document/d/1fxHGnqPGCKFlxfjNw8T3W7r8BgQZlMQhWZNDH9LX3Hk/edit?usp=sharing).

The most recent commit (_Improve UX of universal plugins_) is not covered by the design document. It does a number of changes (detailed in the full commit message) to make the feature as easy-to-use as possible).

### Screenshots

- _Legacy Plugin_ is a legacy (`model=1`, the default) plugin at `<project_dir>/addons/`.
- _Universal Plugin A_ is a universal (`model=2`) plugin at `<project_dir>/addons/` (project-level, available to this project only).
- _Universal Plugin B_ is a universal (`model=2`) plugin at `<godot_config_dir>/addons/` (user-level, available to all projects).
- _Universal Plugin C_ is a universal (`model=2`) plugin packed as a PCK file at `<godot_config_dir>/addons/` (user-level, available to all projects).

## Plugin settings

image

## Plugin creation dialog

image

## Default plugin code

image
(The explanation is added only for `model=2` plugins.)

## File system dock

image

### Browsing resources

image

## Directory dialog

image
---
**This code is generously donated by IMVU.**

### CODE REVIEW

File: editor/plugin_config_dialog.cpp [hunk 1]
```diff
@@ -78,17 +111,30 @@ void PluginConfigDialog::_on_confirmed() {
...
+			if (model >= 2) {
+				source +=
+						TTR("# This plugin uses the new universal model. Things you need to know:\n") +
+						TTR("# - You can place it inside a specific project, at <project_dir>/addons/ or,\n") +
+						TTR("#   to make it user-wide (available for any project you open under your OS user),\n") +
+						TTR("#   put it instead at <user_editor_settings>/addons/.\n") +
+						TTR("#   In either case it will be available in the Godot file system as addons://plugin_name.\n") +
+						TTR("# - To reference resources, you must use that kind of paths. For instance:\n") +
+						"#   var scene = load(\"addons://my_plugin/scene.tscn\")\n"
+						"\n";
```
Comment (YeldhamDev) [inline editor/plugin_config_dialog.cpp]: The comment should respect the 80 characters limit. Also, all other code comments created by the editor aren't translated, so with this the other should probably follow suit then.

**Comment by HeadClot** [comment]
When this gets committed to the main branch does this mean that we will need to update our plugins to account for this change to plugins? Just curious :)

**Comment by vnen** [comment]
> When this gets committed to the main branch does this mean that we will need to update our plugins to account for this change to plugins? Just curious :)

According to the linked design document, old plugins still work the same way.

**Comment by RandomShaper** [comment]
> According to the linked design document, old plugins still work the same way.

That's right. Anyone not ready or willing to use the new model will enjoy complete backwards compatibility.

**Comment by groud** [comment]
I think we should be able to display the scope of a plugin directly in the filesystem dock. That would avoid messing up with user-wide plugins in some case.

**Comment by fire** [comment]
There's CICD failures here https://travis-ci.org/godotengine/godot/builds/640402635

**Comment by RandomShaper** [comment]
> I think we should be able to display the scope of a plugin directly in the filesystem dock. That would avoid messing up with user-wide plugins in some case.

Good idea. Right now the tooltip you get for children of `addons://` gives information about where it is physically.

However, using a different color or folder icon would be nice. The plugin settings dialog is using the accent color for user-level. That "language" could be used there, too.

**Comment by RandomShaper** [comment]
Not much of general interest. Closing.

**Comment by arkology** [comment]
It'll be completely offtopic, but this is how Godot loses great features.

**Comment by Xrayez** [comment]
@RandomShaper I hope a particular implementation will be discussed at [GIP](https://github.com/godotengine/godot-proposals) at least.

I also wonder what's the real reason for closing the PR, was there any discussion regarding this particular implementation? Not all users follow IRC discussions. There's certainly a general interest in this.

**Comment by akien-mga** [comment]
I'm not sure why @RandomShaper's closing comment was so laconic, but to clarify, we discussed this at length at the past Godot Sprint in October 2019, and then again during GodotCon in February 2020. This is a wanted feature but no in its current form, which is needlessly complex.

A new proposal is expected building on this one, which keeps things simple while allowing having global editor plugins.

**Comment by PiCode9560** [comment]
@RandomShaper Do you still have the design document? I'm curious on how this implementation works.

**Comment by RandomShaper** [comment]
Alas, I don't. I can't find it anymore.

**Pull request closed without merging**